core/services/ocr2/plugins/ocr2keeper/evm/mercury_lookup.go

```diff
@@ -0,0 +1,343 @@
+	q := url.Values{
+		ml.feedLabel:  {ml.feeds[index]},
+		ml.queryLabel: {ml.query.String()},
+	}
+	reqUrl := fmt.Sprintf("%s/client?%s", r.mercury.cred.URL, q.Encode())
+	r.lggr.Debugf("MercuryLookup request URL: %s", reqUrl)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, reqUrl, nil)
+	if err != nil {
+		ch <- MercuryBytes{Index: index, Error: err}
+		return
+	}
+
+	ts := time.Now().UTC().UnixMilli()
+	signature := r.generateHMAC(http.MethodGet, "/client?"+q.Encode(), []byte{}, r.mercury.cred.Username, r.mercury.cred.Password, ts)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Authorization", r.mercury.cred.Username)
+	req.Header.Set("X-Authorization-Timestamp", strconv.FormatInt(ts, 10))
+	req.Header.Set("X-Authorization-Signature-SHA256", signature)
+
+	retryErr := retry.Do(
+		func() error {
+			resp, err1 := r.hc.Do(req)
+			if err1 != nil {
+				r.lggr.Errorf("MercuryLookup GET request fails at block %s for upkeep Id %s feed %s: %v", ml.query.String(), upkeepId.String(), ml.feeds[index], err1)
+				return err1
+			}
+			defer resp.Body.Close()
+			body, err1 := io.ReadAll(resp.Body)
+			if err1 != nil {
+				r.lggr.Errorf("MercuryLookup fails to read response body at block %s for upkeep Id %s feed %s: %v", ml.query.String(), upkeepId.String(), ml.feeds[index], err1)
+				return err1
+			}
+
+			if resp.StatusCode == http.StatusNotFound {
+				// there are 2 possible causes for 404: incorrect URL and querying a block where report has not been generated
+				r.lggr.Errorf("MercuryLookup received status code %d at block %s for upkeep Id %s feed %s", resp.StatusCode, ml.query.String(), upkeepId.String(), ml.feeds[index])
+				// return 404 for retry
+				return fmt.Errorf("%d", http.StatusNotFound)
+			} else if resp.StatusCode != http.StatusOK {
+				return fmt.Errorf("MercuryLookup received status code %d at block %s for upkeep Id %s feed %s", resp.StatusCode, ml.query.String(), upkeepId.String(), ml.feeds[index])
+			}
+
+			var m MercuryResponse
+			err1 = json.Unmarshal(body, &m)
+			if err1 != nil {
+				r.lggr.Errorf("MercuryLookup failed to unmarshal body to MercuryResponse at block %s for upkeep Id %s feed %s: %v", ml.query.String(), upkeepId.String(), ml.feeds[index], err1)
+				return err1
+			}
+			blobBytes, err1 := hexutil.Decode(m.ChainlinkBlob)
+			if err1 != nil {
+				r.lggr.Debugf("MercuryLookup failed to decode chainlinkBlob %s at block %s for upkeep Id %s feed %s: %v", m.ChainlinkBlob, ml.query.String(), upkeepId.String(), ml.feeds[index], err1)
+				return err1
+			}
+			ch <- MercuryBytes{Index: index, Bytes: blobBytes}
+			return nil
+		},
+		// only retry when the error is 404 Not Found
+		retry.RetryIf(func(err error) bool {
+			return errors.Is(err, fmt.Errorf("%d", http.StatusNotFound))
+		}),
+		retry.Context(ctx),
+		retry.Delay(RetryDelay),
+		retry.Attempts(TotalAttempt))
+
+	// if all retries fail, it's very likely the feed IDs are incorrect, put into cooldown
```

Comment:
nit. if all retries fail, it's very likely the feed IDs are incorrect or block number is too old or too new, put into cooldown